### PR TITLE
Update the CIS user creation login to look at linked accounts

### PR DESCRIPTION
**Pull request by april**

### CODE REVIEW

File: rules/activate-new-users-in-CIS.js [hunk 1]
```diff
@@ -1,9 +1,12 @@
 function (user, context, callback) {
   const CHANGEAPI_TIMEOUT = 14000; // milliseconds
+  // let MANAGEMENT_API = auth0;
+  const METADATA = context.primaryUserMetadata || user.user_metadata;  // linked acount, or if not linked, then user
   const PERSONAPI_BEARER_TOKEN_REFRESH_AGE = 64800; // 18 hours
   const PERSONAPI_TIMEOUT = 5000;  // milliseconds
   const PUBLISHER_NAME = 'access_provider';
-  const WHITELISTED_CONNECTIONS = ['email', 'firefoxaccounts', 'github', 'google-oauth2', 'oauth2'];
+  const USER_ID = context.primaryUser || user.user_id;  // linked account, or if not linked, then the user account
+  const WHITELISTED_CONNECTIONS = ['email', 'firefoxaccounts', 'github', 'google-oauth2', 'oauth2', 'auth0'];
```
[the remark this answers is not in the document]
Comment (april) [reply rules/activate-new-users-in-CIS.js:8]: Oh whoops, that was an accident. I put that in there for testing and then usually take it out when I commit.

File: rules/activate-new-users-in-CIS.js [hunk 2]
```diff
@@ -1,9 +1,12 @@
 function (user, context, callback) {
   const CHANGEAPI_TIMEOUT = 14000; // milliseconds
+  // let MANAGEMENT_API = auth0;
```
[the remark this answers is not in the document]
Comment (april) [reply rules/activate-new-users-in-CIS.js]: Yup, already had in a follow up commit.  :)

File: rules/activate-new-users-in-CIS.js [hunk 3]
```diff
@@ -1,9 +1,12 @@
...
+  const METADATA = context.primaryUserMetadata || user.user_metadata;  // linked acount, or if not linked, then user
   const PERSONAPI_BEARER_TOKEN_REFRESH_AGE = 64800; // 18 hours
   const PERSONAPI_TIMEOUT = 5000;  // milliseconds
```
[the remark this answers is not in the document]
Comment (april) [reply rules/activate-new-users-in-CIS.js]: Done!

**Review by april — COMMENTED** [review]
Profiles should now merge, @gene1wood.

**Review by gene1wood — COMMENTED** [review]
# New user, first using SSO today long after account linking was enabled

* User logs in with their first identity, say GitHub
* The `user_metadata` for the GitHub profile is empty
* `rules/activate-new-users-in-CIS.js` creates a CIS profile for the GitHub user_id
* The GitHub Auth0 profile's `user_metadata` is updated to contain `existsInCIS`
* User logs out
* User logs in with their second identity, say FxA
* `rules/link-users-by-email-with-metadata.js` begins
  * `user.user_metadata` is empty
  * Line 176
    * Creates an empty object, updates the empty object with the FxA metadata, updates the object now containing the FxA metdata with the GitHub metadata 
    * submits it as a PATCH to Auth0 to overwrite the GitHub profile's user_metadata, which leaves the `existsInCIS` field for the GitHub profile unchanged
  * `context.primaryUser` is set so that the current login has the GitHub user_id not the FxA user_id
  * `context.primaryUserMetadata` is set to the metadata object of the GitHub user, which contains `existsInCIS`
* The `user.user_metadata` is empty
* `rules/activate-new-users-in-CIS.js` runs and sets `USER_ID` to GitHub user_id and `METADATA` to the GitHub user metadata
  * sees that `METADATA.existsInCIS` is true and exits
* User logs out
* User logs in with their third identity, say Passwordless
* `rules/link-users-by-email-with-metadata.js` begins
  * `user.user_metadata` is empty
  * Line 176
    * Creates an empty object, updates the empty object with the Passwordless metadata, updates the object now containing the Passwordless metdata with the GitHub metadata 
    * submits it as a PATCH to Auth0 to overwrite the GitHub profile's user_metadata, which leaves the `existsInCIS` field for the GitHub profile unchanged
  * `context.primaryUser` is set so that the current login has the GitHub user_id not the Passwordless user_id
  * `context.primaryUserMetadata` is set to the metadata object of the GitHub user, which contains `existsInCIS`
* The `user.user_metadata` is empty
* `rules/activate-new-users-in-CIS.js` runs and sets `USER_ID` to GitHub user_id and `METADATA` to the GitHub user metadata
  * sees that `METADATA.existsInCIS` is true and exits
* User logs out
* User logs back in with their second identity, FxA
* `rules/link-users-by-email-with-metadata.js` begins
  * `user.user_metadata` contains `existsInCIS` because the FxA login is linked to the parent GitHub profile which has `existsInCIS`
  * Account linking hits Case 0 and exits
* The `user.user_metadata` contains `existsInCIS`
* `rules/activate-new-users-in-CIS.js` runs and sets `USER_ID` to GitHub user_id as that's the user's current user_id and `METADATA` to the current user metadata
  * sees that `METADATA.existsInCIS` is true and exits


# Old user

* User logs in with Passwordless
* User logs out
* User logs in with Google and gets ratcheted so they can no longer log in with Passwordless
* User logs out
* User logs in with GitHub and gets ratcheted so they can no longer log in with Google or Passwordless
* User logs out
* We enable account linking in 2017
* User logs in with Passwordless
* `rules/link-users-by-email-with-metadata.js` begins
  * `user.user_metadata` is empty
  * Account linking Case 3 is triggered
      * GitHub is chosen as the primary user_id
      * Passwordless is linked into the GitHub user_id
  * Line 176
    * Creates an empty object, updates the empty object with the Passwordless metadata, updates the object now containing the Passwordless metdata with the GitHub metadata 
    * submits it as a PATCH to Auth0 to overwrite the GitHub profile's user_metadata, which leaves the `existsInCIS` field absent the GitHub profile unchanged
  * `context.primaryUser` is set so that the current login has the GitHub user_id not the Passwordless user_id
  * `context.primaryUserMetadata` is set to the metadata object of the GitHub user, which does not contain `existsInCIS`
* The `user.user_metadata` is empty
* `rules/activate-new-users-in-CIS.js` runs and sets `USER_ID` to GitHub user_id and `METADATA` to the GitHub user metadata
  * sees that `METADATA.existsInCIS` is missing
  * Creates a GitHub CIS profile
  * sets the GitHub user metadata to have `existsInCIS` 
* User logs out
* User logs in with GitHub
* `rules/link-users-by-email-with-metadata.js` begins
  * `user.user_metadata` is empty
  * Account linking Case 1 is triggered
      * GitHub is the primary user_id
      * Line 106 sets `context.primaryUserMetadata` to contain existsInCIS
  * since the primary user_id (GitHub) matches the current user_id, we exit account linking

Why set `context.primaryUserMetadata` on line 106? Is it needed?

**Comment by gene1wood** [comment]
Also, I've confirmed that setting an arbitrary value in `context` both persists to subsequent rule executions and does not persist to a subsequent login. I mention this since though it looks like these two variables are either Auth0 variables or ones we came up with, they're actually diferent

* `context.primaryUserMetadata` : This is a variable we set and use in these two rules, but is just us using the `context` object to pass data between rules
* [`context.primaryUser`](https://auth0.com/docs/rules/context-object#properties) : This is an internal Auth0 context variable that's used for account linking

These two variables just look similar but are different in nature.

**Comment by april** [comment]
> Also, I've confirmed that setting an arbitrary value in `context` both persists to subsequent rule executions and does not persist to a subsequent login. I mention this since though it looks like these two variables are either Auth0 variables or ones we came up with, they're actually diferent
> 
>     * `context.primaryUserMetadata` : This is a variable we set and use in these two rules, but is just us using the `context` object to pass data between rules
> 
>     * [`context.primaryUser`](https://auth0.com/docs/rules/context-object#properties) : This is an internal Auth0 context variable that's used for account linking
> 
> 
> These two variables just look similar but are different in nature.

Yes, that's exactly right. I could rename it something else, but it really is just something that is context for the flow of the authentication.

**Comment by april** [comment]
> Why set `context.primaryUserMetadata` on line 106? Is it needed?

Yes, it's used in line 192 to copy the data over to the primary user object.

**Comment by gene1wood** [comment]
> Yes, it's used in line 192 to copy the data over to the primary user object.

Right, but when Case 1 is triggered, code execution never reaches line 192 because lines [158 - 164](https://github.com/mozilla-iam/auth0-deploy/pull/345/files#diff-1eb1b3ecb2f08f71abca44b5b3daf02c98674a05c2e79eb88cfb8037e10909c8R158-R164) cause the rule to exit

In my description above that's where I said

> since the primary user_id (GitHub) matches the current user_id, we exit account linking

**Review by gene1wood — APPROVED** [review]

**Pull request merged**